A Griffin model for 3D heterogeneous LFR assembly

pak1sol opened the pull request:
Referring to Issue #151

pak1sol commented:
Tag @YaqiWang  for your review on the description of Griffin model

eshemon commented:
@YaqiWang @GiudGiud FYI, this is for a NEAMS M2 deliverable due next week which details NEAMS model contributions to VTB this quarter, so a rush review is appreciated, thank you

YaqiWang commented on lfr/heterogeneous_single_assembly_3D/neutronics_standalone/neutronics.i:
```diff
@@ -0,0 +1,1771 @@
+flat_to_flat = 0.153424         # m
+half_asmpitch = ${fparse flat_to_flat / 2 + duct_thickness + asmgap_thickness}
+
+hA = 0.1007 # Lower Core Plate
```
I assume this block is axial heights. It will be good to add a comment like `# axial heights`. Similarly for the size and number of subdivisions next.

YaqiWang commented on lfr/heterogeneous_single_assembly_3D/neutronics_standalone/neutronics.i:
```diff
@@ -0,0 +1,1771 @@
+    NPolar = 2
+    NAzmthl = 3
+    NA = 1
+    sweep_type = asynchronous_parallel_sweeper
```
Suggest to remove this because it is the default value.

YaqiWang commented on lfr/heterogeneous_single_assembly_3D/neutronics_standalone/neutronics.i:
```diff
@@ -0,0 +1,1771 @@
+  power = ${totalpower}
+  power_density_variable = power_density
+  family = MONOMIAL
+  order = CONSTANT
```
If there is no particular reason to use constant monomial for power density, I suggest to remove `family` and `order` here. The code can automatically decide them based on flux variables.

pak1sol replied:
So, when I didn't specify them, the power variable was `Lagrange` (? if I remember correctly) which gave an artificial non-zero power in non-fuel region due to interpolation. As I used very coarse mesh (like just one radial mesh in a helium hole inside the annular fuel shape), this effect was significant. On top of that, I also observed the heat source transferred from Griffin in its sub-app had a non-physical error seemingly caused by a transfer artifact. The use of constant MONOMIAL here resolved all these issues.

pak1sol replied:
This PR is only for a Griffin standalone calculation. The transfer issue happened in the Griffin-Heat conduction multiapp system. I didn't dig into this deeply when I ran into this issue.  

In fact, I used the constant MONOMIAL type for the power variable more for the first reason. I didn't want to have non-physical heat source in non-fuel region transferred to the heat conduction module. The power distribution was obviously wrong and I don't think power need to be interpolated.

YaqiWang replied:
What I was saying is that the target variable in the transfer must be ELEMENTAL as well. If we use constant monomial, we are reducing the accuracy which is typically not a big deal but concerning.

YaqiWang replied:
This is the instruction from the server:
```
If this message is needed, please contact the sender and have them resend the attachment in the allowed format (e.g. filename.inl.zip)
```
@pak1sol can you rename the file and send it again?

YaqiWang commented on lfr/heterogeneous_single_assembly_3D/neutronics_standalone/neutronics.i:
```diff
@@ -0,0 +1,1771 @@
+  [] 
+  [x_coord]
+    family = LAGRANGE
+    order = FIRST
```
You can use subblock here
```
[InitialCondition]
  type = FunctionIC
  function = x
[]
```
to add IC and do not have to have a separate IC block.

pak1sol replied:
So, I see the error message `section '[AuxVariables/x_coord/ICx]' does not have an associated "Action".`
Isn't the below what you've recommended?
```
  [x_coord]
    family = LAGRANGE
    order = FIRST
    [ICx]
      type = FunctionIC
      function = x
    []
  []
```

pak1sol replied:
Oh.... `InitialCondition` is not a user-defined name..

YaqiWang replied:
Right. It has to be that.

YaqiWang commented:
Do you know why the tests are pending? I'd like to check the webpage posted by Civet for reviewing the documentation, but that seems unavailable because the tests have not done yet.

pak1sol commented:
@YaqiWang Thanks for your review. I'll address your comments. For pending tests, I don't think I have control permission for them. @GiudGiud Can you please turn them on? Or can I have control over them?

pak1sol commented on lfr/heterogeneous_single_assembly_3D/neutronics_standalone/neutronics.i:
```diff
@@ -0,0 +1,1771 @@
+
+  cmfd_acceleration = true
+  coarse_element_id = coarse_element_id
+  cmfd_eigen_solver_type = krylovshur
```
`newton` didn't give a converged solution.

YaqiWang replied:
Do you know why? I asking because typically we recommend `newton` because it is faster and we possibly can make it work in the future. I am not so concerned using `krylovshur` here.

pak1sol replied:
I don't know. In fact, I sent this issue to you and @zachmprince via email to ask for your opinion. This problem could be solved very quickly using the `newton` type and couldn't be converged using the `krylovshur` type about more than 1 year ago. This time, however, only the `krylovshur` type converges the solution, with significantly longer time than the previous `newton` type did in the past.

YaqiWang replied:
Can you reply that email to bring it to the top on my list? We might just need to change few default values for the newton solver to make it work. I hope CMFD solve is still a small portion of the entire wall time.

YaqiWang replied:
Oh! Because the email had a zip file attached. INL server filtered it out. The server does send me a notification but I thought it was just a spam... We should share things either with box or with gitlab commits.

YaqiWang replied:
This is the instruction from the server:
```
If this message is needed, please contact the sender and have them resend the attachment in the allowed format (e.g. filename.inl.zip)
```
@pak1sol can you rename the file and send it again?

pak1sol replied:
The result is updated. The computing time is much more reasonable now. I used the Griffin executable built in one of ANL HPC clusters, but I switched to Sawtooth. Not sure about too big difference though.

YaqiWang replied:
MPI distribution can have quite a big impact. But for now, I think this can be resolved. Looks like Newton option is working.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+## Cross Section Generation using MC2-3
+
+It is worth explaining the procedure to generate cross sections using MC2-3 for this model. The axial leakage effect needs to be considered using the MC2-3/TWODANT two-step procedure for the generation of both homogeneous cross sections for non-fuel axial regions and heterogeneous region-wise cross sections for the fuel axial region. At this point of time, the MC2-3/TWODANT procedure is not fully ...
```
`the MC2-3/TWODANT procedure is not fully available in Griffin` => `the MC2-3/TWODANT procedure is not fully available in Griffin due to a licensing issue with TWODANT`. Also please mention MC2-3 is the cross section generation module in Griffin.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+### First step id=xs1st
+
+The first step is to generate a 1041-group `rzmflx` file using TWODANT [!citep](TWODANT) that MC2-3 calls inside using the executable path `c_twodantexe` specified in the `library` block of an MC2-3 input. If TWODANT is not available, PARTISN [!citep](PARTISN) can be used alternatively since it creates a `rzmflx` file. `l_twodant` in the `control` block needs to be set to true. `c_twodant_group=B ...
```
Add a short description on what `rzmflx` file is, such as what it contains, what is its purpose, etc.

YaqiWang commented:
It is a bit difficult for me to review the markdown file without the webpage. I will wait a little bit on the VTB Documentation test. @pak1sol can you work with @GiudGiud to make sure the tests can get properly started?

pak1sol commented on lfr/heterogeneous_single_assembly_3D/neutronics_standalone/neutronics.i:
```diff
@@ -0,0 +1,1771 @@
+  ReflectingBoundary = 'ASSEMBLY_SIDE'
+  [sn]
+    scheme = DFEM-SN
+    family = MONOMIAL
```
OK. I don't recall I tested this input with `L2_LAGRANGE`. Hope this doesn't change the performance result significantly in the document.

moosebuild commented:
Job [VTB Documentation](https://civet.inl.gov/job/1257658/) on 5a3ce6d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/virtual_test_bed/docs/PRs/169/site)

This comment will be updated on new commits.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+*Contact: Hansol Park, hansol.park@anl.gov*
+
+This VTB model provides a high fidelity neutronics model for a representative example of a lead-cooled fast reactor with an annular MOX (UPuO) fuel whose design is based on an early iteration of an LFR-prototype assembly provided by Westinghouse Electric Company, LLC [!citep](WECLFR). The original purpose of this model is to characterize the impact of various sources of uncertainties, such as the ...
```
The first sentence is quite long. I suggest to break it into two right before `whose design is based on...`.

YaqiWang replied:
Similarly the last second sentence can be broken into two before `using a 3D heterogeneous...`. The second part can be `A 3D heterogeneous single assembly steady-state model was used for this study instead of a full core model for simplicity.`

pak1sol replied:
Addressed.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+## LFR model description
+
+The key parameters of the LFR model are listed in [lfrkeyparam]. The power output is 950 MWth (~450 MWe) for the whole core and the nominal power of an inner core fuel assembly is 3.7 MWth. The fuel assembly pitch is 16.3 cm with a 4 mm lead-filled gap between assemblies and 3.5 cm thick duct wall as shown on the right in [lfrgeom]. Each assembly contains 127 cladded fuel pins arranged in a trian ...
```
I think it is more accurate to differentiate between model and design. For example, it is better to say `the LFR design is described`.

pak1sol replied:
The head `LFR model description` is updated to `Description of LFR design`.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+## Cross Section Generation using MC2-3
+
+It is worth explaining the procedure to generate cross sections using MC2-3 for this model. MC2-3 is one of the cross section generation modules in Griffin. The axial leakage effect needs to be considered using the MC2-3/TWODANT two-step procedure for the generation of both homogeneous cross sections for non-fuel axial regions and heterogeneous region-wise cross sections for the fuel axial region ...
```
`MC2-3 is one of the cross section generation modules in Griffin.` to `MC2-3, the cross section generation module, is one of the modules in Griffin.` or just `is the cross section generation module of Griffin`.

pak1sol replied:
Addressed.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+Homogenized compositions representing different axial regions are specified in the `material` block. In this step, temperature does not need to be very accurate.
+
+!listing lfr/heterogeneous_single_assembly_3D/cross_section/Step1/LFR_InnerFuel_HOM_1041g_TWODANT.mcc3.sh start=\$material end=/ include-end=True
```
`start` and `end` here seem not providing the right section of the bash script.

pak1sol replied:
`start` and `end` are now properly set. I changed `Homogenized compositions` to `Volume-homogenized compositions`. I think this would suffice to explain how those atomic densities are calculated. Basically, you are correct.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+!listing lfr/heterogeneous_single_assembly_3D/cross_section/Step1/LFR_InnerFuel_HOM_1041g_TWODANT.mcc3.sh start=\$material end=/ include-end=True
+
+The `twodant` block is to generate a TWODANT input internally for calling TWODANT. `niso` and `ngroup` are filled by MC2-3, `mt` and `nzone` are the number of materials and coarse meshes, `im` and `jm` are the number of coarse meshes in the x and y directions, `it` and `jt` are the total number of fine meshes in the x and y directions, `igeom=6` means x-y geometry, `isn` is the Sn order, and `max ...
```
Is memory size in bytes?

pak1sol replied:
I think the size is in a unit of `word` = 2 bytes. I'll specify this.

pak1sol replied:
I added `(in a unit of word = 2 bytes)`.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+Input parameters from `isct` to `rmflux` are solver options. Important ones are `isct`: Legendre order of scattering, `ievt=1`: calculation type=k-eff, `ibl, ibr, ibt, ibb`: boundary conditions for left, right, top and bottom (0/1/2/3=vacuum/reflective/periodic(only for top and bottom)/white), `epsi`: convergence precision, `oitm`: maximum number of outer iterations, `iitm, iitl`: maximum number  ...
+
+!listing lfr/heterogeneous_single_assembly_3D/cross_section/Step1/LFR_InnerFuel_HOM_1041g_TWODANT.mcc3.sh start=\$twodant end=/ include-end=True
```
This start/end has issue as well.

YaqiWang replied:
Will not comment on this anymore. Suggest to check the link posted by civet to make sure all lists have the proper start and end.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+### Second step id=xs2nd
+
+The second step is to condense the UFG cross section to the broad group (BG) cross section using the TWODANT spectrum. Cross sections for the activel fuel region and the other regions are prepared separately and merged at the last step.
```
Here is UFG 1041 or 2082?

pak1sol replied:
It's 2082 groups.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+### First step id=xs1st
+
+The first step is to generate a 1041-group `rzmflx`, which is the flux solution file of a TWODANT [!citep](TWODANT) SN calculation. It contains zone-wise Legendre moments of the angular flux solution up to a specified order. TWODANT is called inside MC2-3 when its executable path is specified by the `c_twodantexe` input parameter in the `library` block of an MC2-3 input with `l_twodant=T`. If TWO ...
```
What is the input data for this TWODANT calculation? The 1041-group library?

pak1sol replied:
The ISOTXS output file of the first MC2-3 calculation is the TWODANT input for cross sections. Let me add this description.

pak1sol replied:
I added more detailed description on the composition specification in the MC2-3 input. I also added the following paragraph.

```
The cross section set for the TWODANT calculation is the ISOTXS output file of the MC2-3 calculation in this first step. MC2-3 fills the matls option in the twodant block for the native TWODANT input with isotope names and atomic number densities taken from the material block. Thus, users do not need to take care of this part.
```

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+       id=onedxs_model
+       caption=1D R-geometry model for hexagonal ring-wise cross section generation (red: fuel + helium, blue: lead, gray: clad, cyan: duct)
+
+Below shows the `control` block. Specifying `c_geometry_type=cylinder` automatically sets the problem type as a 1D CPM transport problem. `i_number_region` means the number of different cross section regions. The self-shielding condition is calculated per cross section region and thus UFG cross sections are unique per cross section region. `c_externalspectrum_ufg=rzmflx` means to use the TWODANT  ...
```
I am confused by this second step. You are performing 1D transport calculation. Looks like you are not using the UFG group structure for this calculation. Then how many groups in this transport calculation? What is that `ANL9` group structure? You mentioned later the TWODANT spectrum is superimposed to the region-wise spectrum, how is the superimposition is done?

pak1sol replied:
I added the following for clarification.

```
For clarification, the TWODANT calculation in the first step is the 1041 group calculation. In this secton step, the 0D UFG calculation is performed for each mixture of non-fuel axial regions and the 1D UFG calculation is performed for the fuel axial region by MC2-3. Then, resulting UFG spectra are superimposed to 1041-group TWODANT spectra for condensation of UFG cross sections to the broad group structure specified in the c_group_structure of the &control block.
```

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+### First step id=xs1st
+
+The first step is to generate a 1041-group `rzmflx`, which is the flux solution file of a TWODANT [!citep](TWODANT) SN calculation. It contains zone-wise Legendre moments of the angular flux solution up to a specified order. TWODANT is called inside MC2-3 when its executable path is specified by the `c_twodantexe` input parameter in the `library` block of an MC2-3 input with `l_twodant=T`. If TWO ...
```
Is this a 2D or 1D calculation? and what are the boundary conditions? How do you specify the boundary conditions?

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+The `twodant` block is to generate a TWODANT input internally for calling TWODANT. `niso` and `ngroup` are filled by MC2-3, `mt` and `nzone` are the number of materials and coarse meshes, `im` and `jm` are the number of coarse meshes in the x and y directions, `it` and `jt` are the total number of fine meshes in the x and y directions, `igeom=6` means x-y geometry, `isn` is the Sn order, and `max ...
+
+Input parameters from `isct` to `rmflux` are solver options. Important ones are `isct`: Legendre order of scattering, `ievt=1`: calculation type=k-eff, `ibl, ibr, ibt, ibb`: boundary conditions for left, right, top and bottom (0/1/2/3=vacuum/reflective/periodic(only for top and bottom)/white), `epsi`: convergence precision, `oitm`: maximum number of outer iterations, `iitm, iitl`: maximum number  ...
```
I think here you need to mention that left/right boundary are reflecting, thus this 2D calculation is a pseudo 1D problem. The reason of doing this is that TWODANT only supports 2D calculations. I was confused on this when I saw 1D slab model in Figure 6 because I was not paying attention on the boundary condition here.

pak1sol replied:
I added `epso`.
```
epsi: inner iteration convergence precision (default=0.0001), epso: outer iteration convergence precision (default=epsi),
```
To avoid confusion, I added the following sentence in the `First step`.
```
In this problem, a pseudo one-dimensional (1D) slab problem is solved using TWODANT by assigning the reflective boundary condition on the left and right boundaries in the x-y geometry. 
```

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
````diff
@@ -0,0 +1,277 @@
+The order to run two scripts doesn't matter. Once the above two scripts are run, `LFR_127Pin_Fuel_1D_9g.mcc3.sh.ISOTXS` and `LFR_127Pin_NonFuel_9g.mcc3.sh.ISOTXS` files are created. The next step is to convert those two ISOTXS files to isoxml files by running
+
+```
+isoxml-opt -n ./Step2/Fuel/LFR_127Pin_Fuel_1D_9g.mcc3.sh.ISOTXS
````
Maybe it is better to replace `isoxml-opt -n` with `griffin-opt --isoxml--nisotxs` now because it is a bit hard to explain what `isoxml-opt` is.

pak1sol replied:
Accepted.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+!listing lfr/heterogeneous_single_assembly_3D/cross_section/Step2/NonFuel/LFR_127Pin_NonFuel_9g.mcc3.sh.xml start=<Multigroup_Cross_Section_Library end=<Tabulation> include-end=False
+
+Without this manual change of the library ID, two xml files cannot be merged. Finally, two xml files are merged using the merging capability of ISOXML to produce the final xml file as
```
Is there a way to populate the ID in ISOXML properly? Like if ISOTXS stores an ID, that ID can be used by the converter, or we can add a command-line option like `--isoxml-isotxs-id` to assign the ID when converting. This does not have to be in place here.

pak1sol replied:
Unfortunately, there are limitations to populate the ID in ISOXML as intended. Regarding this, I made a major change in the MC2-3 script and the Griffin input as well as the library. I added detailed description on ISOTXS->ISOXML conversion part and isoxml merging process.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+### Input parameters
+
+Defining input parameters in advance is beneficial to make the input tractable. Radial and axial specifications of pin and assembly, block IDs and material IDs to be assigned to block IDs and the total power are defined as input parameters to be used in the main input later.
```
add `because the same value can be used in multiple places` at the end of the first sentence.

pak1sol replied:
Added.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+The mesh system is built on-the-fly in a one input using the MOOSE reactor module [!citep](MOOSEReactorModule). 
+
+First, pins for each hexagonal ring are defined using `PolygonConcentricCircleMeshGenerator`. Since there are 7 rings, 7 pins are defined. They have different block IDs (`bid_F_fuel_R1` to `bid_F_fuel_R7`) for fuel region so that different material IDs can be assigned later. Since different element types cannot have the same block ID, the innermost helium hole of a triangular element and the heli ...
```
`different element types cannot have the same block ID` is not exactly right. It is a rule imposed by Exodus mesh format. So I think it is better to rephrase this as `generally different element types should have different block ID` to weaken the words a bit here.

pak1sol replied:
Accepted.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+!listing lfr/heterogeneous_single_assembly_3D/neutronics_standalone/neutronics.i start=[assign] end=[] include-end=True
+
+A coarse mesh is generated for the CMFD acceleration. For better alignment of peripheral pin meshes near duct, only the peripheral pins have two radial meshes divided by the cladding outer radius while inner pins have one radial mesh. Azimuthal divisions are the same as the fine mesh. The duct mesh is also the same as the fine mesh, otherwise the CMFD solve didn't converge. The axial mesh is also ...
```
Coarser duct mesh can cause CMFD divergence? You mean CMFD solver diverges but not Richardsion iteration, right?

pak1sol replied:
CMFD solve diverged, not Richardson iteration.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+### Executioner
+
+The `SweepUpdate` executioner is used for the CMFD acceleration. `SweepUpdate` is a special Richardson executioner for performing source iteration with a transport sweeper. This source iteration can be accelerated by turning on `cmfd_acceleration` which invokes the CMFD solve where the low order diffusion equation is solved using transport-corrected Diffusion coefficients. `coarse_element_id` is  ...
```
`using transport-corrected Diffusion coefficients` is not quite right, should be replace with `with a convection closure term to make the diffusion system and the transport system consistent`.

YaqiWang replied:
remove `to correct Diffusion coefficients`.

pak1sol replied:
Accepted.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+The `SweepUpdate` executioner is used for the CMFD acceleration. `SweepUpdate` is a special Richardson executioner for performing source iteration with a transport sweeper. This source iteration can be accelerated by turning on `cmfd_acceleration` which invokes the CMFD solve where the low order diffusion equation is solved using transport-corrected Diffusion coefficients. `coarse_element_id` is  ...
+
+`Richardson_rel_tol` or `Richardson_abs_tol` is the tolerance used to check the convergence of the Richardson iteration. If `richardson_postprocessor` is specified, its `PostProcessor` value is used as the convergence metric. Otherwise Griffin uses the L2 norm difference of the angular flux solution between successive Richardson iterations, which is added to `richardson_postprocessor` with the na ...
```
`Richardson_rel_tol` is +the minimum value+ for, should the minimum value be the tolerance? Also `R` needs to be replaced with `r`. There are several other places that should have used lower case r.

YaqiWang replied:
than `none` does => than `none`.

pak1sol replied:
Accepted.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+### Flux normalization for producing the total power
+
+The flux solution can be normalized for producing the total power using the `[PowerDensity]` block. `power` is the user-specified total power and `power_density_variable` is a required `AuxVariable` name for power density. Power is calculated using the `kappaFission` cross section in the cross section library which is the total recoverable energy per fission multiplied by the fission cross sectio ...
```
Why not account for non-fission heating? We have a control parameter `legacy_energy_deposition_definition`. Oh, looks like `[PowerDensity]` block does not have this.

YaqiWang replied:
Suggest to replace this with a `!alert note` and add that this will be added in the future.

pak1sol replied:
`!alert note` is added.

YaqiWang commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+[CMFDperformance] compares performance of calculations with and without CMFD using different input options using 144 processors in BEBOP high performance computing clusters of Argonne National Laboratory. The top left plot shows the total number of Richardson iterations until convergence with the relative tolerance of $10^{-4}$. The legend applies to the other three plots, too. The top right plot ...
+
+First of all, from the top left plot, the `GMRes` inner solve type is very effective, immediately dropping the number of Richardson iterations from $39$ to $14$. Then, the number of Richardson iterations does not decrease much over increasing maximum number of inner iterations, which means the small number of inner iterations is sufficient for `GMRes`. Meanwhile, `SI` is not immediately effective ...
```
SI gives smaller number of Richardson iterations than GMRES with large number max inner iteration possibly because of the reflecting boundary condition, where two methods have different treatments. But this needs to have numerical proof so we do not have to include it here.

YaqiWang replied:
CMFD time is pretty bad. Not sure if this is expected. Do you have the inputs? I want to run it quickly on Sawtooth to see if I can observe the same thing. Maybe @zachmprince is helping already?

pak1sol replied:
This part is updated. Now the result looks pretty good. Zach helped me. CMFD speed-up is about 10 times and this is explicitly mentioned.

YaqiWang commented:
@pak1sol I have difficulties on understanding the MC2-3 parts especially the second step. Hopefully after you address my comments, things become clearer to me. This is a great work, an exact tutorial we need for FR. BTW, are we going to include NekRS and the multiphysics part in the future?

GiudGiud commented:
I would rather we did not. It s better to refer to our papers that are out in the public.

eshemon commented:
> @pak1sol I have difficulties on understanding the MC2-3 parts especially the second step. Hopefully after you address my comments, things become clearer to me. This is a great work, an exact tutorial we need for FR. BTW, are we going to include NekRS and the multiphysics part in the future?

Thanks Yaqi - yes, we are working on the multiphysics part with NekRS (Cardinal) + MOOSE Heat Conduction. We have preliminary coupling for single and 7-pin. Once these are fully working we can scale up to the full assembly working later this year.

YaqiWang commented:
> I would rather we did not. It s better to refer to our papers that are out in the public.

I might not make myself clear. I mean using `[SweepUpdate.md]` in the markdown file here. Because Griffin is controlled, we probably cannot do that.

GiudGiud commented:
Yes we cannot do that.

pak1sol commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+For the fuel region, fine mesh-based region-wise cross sections need to be generated as the Griffin transport calculation is fine-mesh based. This can be achieved by setting up an MC2-3 input that employs a 1-dimensional (1D) heterogeneous tranport calculation using the collision probability method (CPM) whose model is shown in [onedxs_model]. Hexagonal ring-wise cross sections are generated for  ...
+
+!media lfr/onedxs_model.png
```
I made the figure larger.

pak1sol commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+       id=XS_Hom_Vs_Reg
+       caption=Relative difference of assembly average cross section to hexagonal ring-wise fuel cross section for U-238 capture
+
+For the non-fuel region, the setting is almost similar to the first step input except `c_externalspectrum_ufg=rzmflx` instead of `l_twodant=T` in the `control` block and `i_externalspectrum(composition #)=TWODANT zone #` in the `material` block.
```
`l_twodant=T` is not in the script given in this part, but in the script in the `First step`.

pak1sol commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,277 @@
+
+!listing lfr/heterogeneous_single_assembly_3D/neutronics_standalone/neutronics.i start=This assembles the 7 pins into an assembly with a duct end=[] include-end=True
+
+Before extruding the 2D assembly geometry built above, a name of the duct and lead interface is explicitly defined.
```
Accepted.

pak1sol commented:
I think I addressed all your comments @YaqiWang. Please check them. And I noted I forgot including the description of the `[Materials]` block, so I added the part. You can check that out.
Thank you for your thorough review!

pak1sol commented:
> @pak1sol Thanks for addressing my comments. Now things are clear to me. I only have one question left: how did you prepare the atomic density lists in MC2-3 inputs? You have your own spreadsheet for evaluating them based on design description?
-> Yes, exactly.

> Another note: maybe we should add a future work section? In there we can mention that we are streamlining the calculation and we can use one single Griffin input to generate xs, and perform full-core transport calculation. Multiphysics coupling can also be mentioned.
-> OK. Let me add this part. Thank you!

GiudGiud commented:
The testing fails, the neutron library isnt checked in the repo.
Did you mean to use LFS for it?
```
data/civet1/build/virtual_test_bed/lfr/heterogeneous_single_assembly_3D/neutronics_standalone.lfr_neutronics_alone: Cannot find the multigroup library with ID 13 of library ISOTXS-neutron in file /data/civet1/build/virtual_test_bed/lfr/heterogeneous_single_assembly_3D/neutronics_standalone/../cross_section/LFR_127Pin_9g.xml
```

GiudGiud commented:
Actually it does seem checked in.

GiudGiud commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,298 @@
+
+In the `material` block, 7 different materials are defined for the fuel regions of 7 hexagonal rings. Even though the composition is the same, different isotope names are assigned to the same isotope to generate cross sections separately. On the other hand, only a single material is defined for each of lead, clad and duct regions, which means that only the average values over multiple cross secti ...
+
+Suffices in all compositions need special attention since it is directly related to how the ISOTXS file will be converted to the ISOXML file. When the resulting ISOTXS file that contains all isotopes in all compositions without composition IDs is converted to the ISOXML file, the last single character in the user-defined name is used to differentiate compositions and store cross sections of diffe ...
```
```suggestion
Suffices in all compositions need special attention since it is directly related to how the ISOTXS file will be converted to the ISOXML file. When the resulting ISOTXS file that contains all isotopes in all compositions without composition IDs is converted to the ISOXML file, the last single character in the user-defined name is used to differentiate compositions and store cross sections of different compositions in different library IDs. Library IDs are increasing in the alphabetical order of the last single character, from the starting value of $1$. In this MC2-3 input, eleven compositions are differentiated by the characters from A to K and the library IDs of them in the ISOXML file to be converted in a later stage are from $1$ to $11$. If a isotope name excluding the last single character is not the standard isotope name defined in ISOXML, the pre-fix `pseudo_` is attached. Otherwise, the isotope name is used as is.
```

GiudGiud commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,298 @@
+
+The key parameters of the LFR design are listed in [lfrkeyparam]. The power output is 950 MWth (~450 MWe) for the whole core and the nominal power of an inner core fuel assembly is 3.7 MWth. The fuel assembly pitch is 16.3 cm with a 4 mm lead-filled gap between assemblies and 3.5 cm thick duct wall as shown on the right in [lfrgeom]. Each assembly contains 127 cladded fuel pins arranged in a tria ...
+
+The active fuel region in the middle is located between upper and lower thermal insulators, gas planum, bundle grid and pins plug, and inlet and outlet wrappers. Bundle grid and pins plug, inlet and outlet wrappers, and lower core plate are radially homogenized while radially heterogeneous fine-meshes were used in the other five axial regions. The annular fuel shape in [annularfuel] is explicitly ...
```
```suggestion
The active fuel region in the middle is located between upper and lower thermal insulators, gas plenum, bundle grid and pins plug, and inlet and outlet wrappers. Bundle grid and pins plug, inlet and outlet wrappers, and lower core plate are radially homogenized while radially heterogeneous fine-meshes were used in the other five axial regions. The annular fuel shape in [annularfuel] is explicitly modeled for the active fuel region and materials inside the cladding are radially homogenized for the four axial regions of thermal insulators and gas plenum.
```
correct elsewhere too pls

GiudGiud commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,298 @@
+       id=MCCgenProc
+       caption=MC2-3 cross section generation procedure with the axial leakage effect considered using TWODANT
+
+Without `l_spatial_homogenization=F` in an input, which is easy to make mistake, one will end up using the assembly average cross section for each region in heterogeneous fine-mesh transport calculation. [XS_Hom_Vs_Reg] shows the relative difference of assembly average cross section to hexagonal ring-wise fuel cross section for U-238 capture. These differences are basically the same as the spatia ...
```
```suggestion
Without `l_spatial_homogenization=F` in an input, which is an easy mistake to make, one will end up using the assembly average cross section for each region in heterogeneous fine-mesh transport calculation. [XS_Hom_Vs_Reg] shows the relative difference of assembly average cross section to hexagonal ring-wise fuel cross section for U-238 capture. These differences are basically the same as the spatial self-shielding factor of each ring: the ratio of hexagonal ring-wise flux to assembly average flux. Since spectrum is softer for peripheral regions than for central regions, central region flux is lower than the assembly average flux at low energy and higher at high energy. Thus, a test showed that the use of assembly average cross section instead of region-wise cross section resulted in about +400 pcm error in k-effective due to underestimation of low energy capture and overestimation of high energy fission.
```

GiudGiud commented on doc/content/lfr/heterogeneous_single_assembly_3D/Griffin_standalone_LFR.md:
```diff
@@ -0,0 +1,298 @@
+!listing lfr/heterogeneous_single_assembly_3D/neutronics_standalone/neutronics.i start=[PowerDensity] end=[] include-end=True
+
+!alert note
+Non-fission heating will be allowed in the `[PowerDensity]` block in the future.
```
```suggestion
Non-fission heating will be added in the `[PowerDensity]` block in the future.
```

GiudGiud reviewed (commented):
@YaqiWang or @mtnwalker can you look at the test failure? 
https://civet.inl.gov/job/1264280/

It doesn't make sense that it cant find that ID to me? The id is present in the file that is checked in

I can fix the text things

YaqiWang commented:
The Griffin module needs an update. We need a version later than Dec. 9th but the version in the test seems on Nov. 23.

GiudGiud commented on lfr/heterogeneous_single_assembly_3D/neutronics_standalone/tests:
```diff
@@ -0,0 +1,19 @@
+    cli_args = "Executioner/richardson_max_its=1"
+    executable_pattern = 'griffin*|blue_crab*'
+    prereq = 'lfr_neutronics_syntax'
+  []
```
```suggestion
    skip = '#176'
  []
```

GiudGiud commented:
BlueCrab failures are unrelated.

GiudGiud commented:
Thanks for the contribution and for looking into the issues @pak1sol @zachmprince @YaqiWang 
And thanks for reviewing @YaqiWang !

YaqiWang commented:
Glad to help and thank you! @GiudGiud

pak1sol commented:
@GiudGiud I'm sorry I was absent for a personal reason for a time being. Thank you for checking and merging this.

GiudGiud commented:
Emily told me. Congratulations :)
